src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -1385,6 +1385,11 @@ private static LambdaConversionResult IsAnonymousFunctionCompatibleWithExpressio
                 return LambdaConversionResult.ExpressionTreeFromAnonymousMethod;
             }
 
+            if (delegateType is null)
```

Comment:
nit: Consider `if (type.Arity == 0) ...` if that is more directly what is intended, and assert that `delegateType` is not null otherwise.

Reply:
I'll leave as is, since `delegateType` here means the type argument to `Expression<TDelegate>` which is `null` for non-generic `Expression`.